File: code/graph-algorithms/cycle_undirected_graph/CycleUndirectedGraph.cs
// Part of Cosmos by OpenGenus Foundation

using System;
using System.Collections.Generic;

namespace CycleUndirectedGraph
{
    class Graph
    {
        protected Dictionary<int, HashSet<int>> adjList;

        public Graph()
        {
            adjList = new Dictionary<int, HashSet<int>>();
        }

        /// <summary>
        /// Adds an edge from node1 to node2 in the graph
        /// </summary>
        public void addEdge(int node1, int node2)
        {
            if (!adjList.ContainsKey(node1))
                adjList.Add(node1, new HashSet<int>());
            if (!adjList.ContainsKey(node2))
                adjList.Add(node2, new HashSet<int>());

            adjList[node1].Add(node2);
            adjList[node2].Add(node1);
        }

        private bool cyclicHelper(int current, Dictionary<int, bool> visited)
        {
            visited[current] = true;

            foreach(int node in adjList[current])
            {
                if (!visited[node])
                    if (cyclicHelper(node, visited, current))
                        return true;
            }

            return false;
        }

        private bool cyclicHelper(int current, Dictionary<int, bool> visited, int parent)
        {
            visited[current] = true;

            foreach (int node in adjList[current])
            {
                if (!visited[node])
                {
                    if (cyclicHelper(node, visited, current))
                        return true;
                }
                else if (node != parent)
                    return true;
            }

            return false;
        }

        /// <summary>
        /// Returns true if the graph contains a cycle, false otherwise
        /// </summary>
        public bool isCyclic()
        {
            if (adjList.Count == 0)
                return false;

            Dictionary<int, bool> visitDict = new Dictionary<int, bool>();

            foreach (int key in adjList.Keys)
                visitDict.Add(key, false);

            List<int> keyListCopy = new List<int>(visitDict.Keys);

            foreach (int key in keyListCopy)
            {
                if (!visitDict[key])
                    if (cyclicHelper(key, visitDict))
                        return true;
            }

            return false;
        }
    }

    // Testing with this graph first:
    // 1 - 0 - 3
    // |  /    |
    // 2       4
    // Followed by this graph:
    // 0 - 1 - 2

    // Expected Output:
    // Test 1:
    // True
    // Test 2:
    // False
    class UndirectedCycleTest
    {
        static int Main(string[] args)
        {
            Graph testGraph = new Graph();
            testGraph.addEdge(1, 0);
            testGraph.addEdge(1, 2);
            testGraph.addEdge(2, 0);
            testGraph.addEdge(0, 3);
            testGraph.addEdge(3, 4);
            Console.Out.WriteLine("Test 1:");
            Console.Out.WriteLine(testGraph.isCyclic());

            testGraph = new Graph();
            testGraph.addEdge(0, 1);
            testGraph.addEdge(1, 2);
            Console.Out.WriteLine("Test 2:");
            Console.Out.WriteLine(testGraph.isCyclic());
            return 0;
        }
    }
}

File: code/graph_algorithms/src/cycle_undirected_graph/cycle_undirected_graph.cpp
/*
 * Part of Cosmos by OpenGenus Foundation
 */
#include <iostream>
#include <list>
#include <limits.h>
using namespace std;

class Graph
{
    int V;    // No. of vertices
    list<int> *adj;    // Pointer to an array containing adjacency lists
    bool isCyclicUtil(int v, bool visited[], int parent);
public:
    Graph(int V);
    void addEdge(int v, int w);
    bool isCyclic();   // returns true if there is a cycle
};

Graph::Graph(int V)
{
    this->V = V;
    adj = new list<int>[V];
}

void Graph::addEdge(int v, int w)
{
    adj[v].push_back(w);  //Undirected, so adding edge to both lists
    adj[w].push_back(v);
}

bool Graph::isCyclicUtil(int v, bool visited[], int parent)
{
    visited[v] = true;
    list<int>::iterator i;
    for (i = adj[v].begin(); i != adj[v].end(); ++i)
    {
        if (!visited[*i])
        {
            if (isCyclicUtil(*i, visited, v))
                return true;
        }
        else if (*i != parent)
            return true;
    }
    return false;
}

bool Graph::isCyclic()
{
    bool *visited = new bool[V];
    for (int i = 0; i < V; i++)
        visited[i] = false;

    for (int u = 0; u < V; u++)
        if (!visited[u])
            if (isCyclicUtil(u, visited, -1))
                return true;
    return false;
}

int main()
{
    Graph g1(5);
    g1.addEdge(1, 0);
    g1.addEdge(0, 2);
    g1.addEdge(2, 0);
    g1.addEdge(0, 3);
    g1.addEdge(3, 4);
    g1.isCyclic() ? cout << "Graph contains cycle\n" :
        cout << "Graph doesn't contain cycle\n";

    Graph g2(3);
    g2.addEdge(0, 1);
    g2.addEdge(1, 2);
    g2.isCyclic() ? cout << "Graph contains cycle\n" :
        cout << "Graph doesn't contain cycle\n";

    return 0;
}
